File: backend/src/main/java/com/slpolice/trafficfines/fine/dto/FineDTO.java
package com.slpolice.trafficfines.fine.dto;

import com.slpolice.trafficfines.fine.entity.Fine;
import com.slpolice.trafficfines.fine.entity.FineStatus;
import lombok.Builder;
import lombok.Data;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * Data Transfer Object for a Fine — used in all API responses.
 * As per the DTO Pattern: "Never expose JPA entities directly."
 * This DTO is the agreed API contract for the fine payload shared with
 * the frontend teams (Member 3, 4) and documented in the parallel_development_guide.
 */
@Data
@Builder
public class FineDTO {

    private Long id;
    private String referenceNumber;

    // Officer info
    private Long officerId;
    private String officerName;
    private String officerDistrict;

    // Category info
    private String categoryCode;
    private String categoryDescription;
    private BigDecimal amount;

    // Driver info
    private String driverLicenseNo;
    private String driverName;
    private String vehicleNumber;
    private String district;
    private String location;

    private FineStatus status;
    private LocalDateTime issuedAt;  // maps to createdAt

    /**
     * Static factory method to convert a Fine entity to a FineDTO.
     * Keeps mapping logic in one place (avoid scattering it across service/controller).
     */
    public static FineDTO from(Fine fine) {
        return FineDTO.builder()
                .id(fine.getId())
                .referenceNumber(fine.getReferenceNumber())
                .officerId(fine.getOfficer().getId())
                .officerName(fine.getOfficer().getFullName())
                .officerDistrict(fine.getOfficer().getDistrict())
                .categoryCode(fine.getCategory().getCategoryCode())
                .categoryDescription(fine.getCategory().getDescription())
                .amount(fine.getCategory().getAmount())
                .driverLicenseNo(fine.getDriverLicenseNo())
                .driverName(fine.getDriverName())
                .vehicleNumber(fine.getVehicleNumber())
                .district(fine.getDistrict())
                .location(fine.getLocation())
                .status(fine.getStatus())
                .issuedAt(fine.getCreatedAt())
                .build();
    }
}

File: mobile/lib/data/models/fine_dto.dart
class FineDTO {
  final int id;
  final String referenceNumber;
  final int officerId;
  final String officerName;
  final String officerDistrict;
  final String categoryCode;
  final String categoryDescription;
  final double amount;
  final String driverLicenseNo;
  final String driverName;
  final String vehicleNumber;
  final String district;
  final String location;
  final String status;
  final String issuedAt;

  FineDTO({
    required this.id,
    required this.referenceNumber,
    required this.officerId,
    required this.officerName,
    required this.officerDistrict,
    required this.categoryCode,
    required this.categoryDescription,
    required this.amount,
    required this.driverLicenseNo,
    required this.driverName,
    required this.vehicleNumber,
    required this.district,
    required this.location,
    required this.status,
    required this.issuedAt,
  });

  factory FineDTO.fromJson(Map<String, dynamic> json) {
    return FineDTO(
      id: json['id'] ?? 0,
      referenceNumber: json['referenceNumber'] ?? '',
      officerId: json['officerId'] ?? 0,
      officerName: json['officerName'] ?? '',
      officerDistrict: json['officerDistrict'] ?? '',
      categoryCode: json['categoryCode'] ?? '',
      categoryDescription: json['categoryDescription'] ?? '',
      amount: (json['amount'] ?? 0).toDouble(),
      driverLicenseNo: json['driverLicenseNo'] ?? '',
      driverName: json['driverName'] ?? '',
      vehicleNumber: json['vehicleNumber'] ?? '',
      district: json['district'] ?? '',
      location: json['location'] ?? '',
      status: json['status'] ?? '',
      issuedAt: json['issuedAt'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'referenceNumber': referenceNumber,
      'officerId': officerId,
      'officerName': officerName,
      'officerDistrict': officerDistrict,
      'categoryCode': categoryCode,
      'categoryDescription': categoryDescription,
      'amount': amount,
      'driverLicenseNo': driverLicenseNo,
      'driverName': driverName,
      'vehicleNumber': vehicleNumber,
      'district': district,
      'location': location,
      'status': status,
      'issuedAt': issuedAt,
    };
  }
}
